### PR TITLE
m 3741 integrate slow query log package to mmc 1

### PR DESCRIPTION
**Pull request by Or92**
- tested and found working
- tested
- tested
- tested
- tested
- tested
- tested

### CODE REVIEW

File: client-agent/src/utils/scheduled-job.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    async start() {
+        try {
+            const start = process.hrtime();
+            const results = await this.task();
+            const end = process.hrtime(start);
+            logger.info(`task execution took ${end[0]} sec`);
+
+            let ms;
```
Comment (gidimetis) [inline client-agent/src/utils/scheduled-job.js:21]: let ms = (end[0] < (this.PERIOD_IN_MINUTES * 60)) ?   ((this.PERIOD_IN_MINUTES * 60) - end[0]) * 1000 : 0 ;

File: client-agent/src/slow-query-log.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+function getQueries(fakeHoursDelta) {
+    const now = new Date();
+    now.setHours(now.getHours() - fakeHoursDelta);
+    const currentMinutes = now.getMinutes();
+    const currentHour = IGNORE_CURRENT_TIME ? 0 : now.getHours();
+    if (process.argv.length === 2) {
+        return Object.keys(QUERIES)
+            .filter((key) => relevant(QUERIES[key].times_a_day, currentHour, currentMinutes))
```
Comment (gidimetis) [inline client-agent/src/slow-query-log.js:50]: reduce instead filter and map ?

File: client-agent/src/slow-query-log.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        return;
+    }
+
+    const promiseArr = connections.map(
+        (connection) => {
+            const { connectionString, database } = connection;
+            return takeAction(connectionString, database);
+        }
+    );
+    logger.info('Collection is done.');
```
Comment (gidimetis) [inline client-agent/src/slow-query-log.js:81]:
```suggestion
    logger.info('Slow query log Collection is done.');
```

File: client-agent/src/collect-runner.js [hunk 4]
```diff
@@ -1,50 +1,78 @@
...
-  catch (error) {
-    logger.error('Runner cant continue: ', { error });
-    process.exit(1);
-  }
-})()
+let connections;
 
 async function app(hostedOnAws) {
+  logger.info('app is staring');
   return setup()
```
Comment (gidimetis) [inline client-agent/src/collect-runner.js:15]: there are 4 nested promises here , maybe simplified it ?

File: client-agent/src/utils/scheduled-job.js (hunk 1 above)
Comment (gidimetis) [inline client-agent/src/utils/scheduled-job.js:21]: let ms =  (end[0] < (this.PERIOD_IN_MINUTES * 60)) ? ((this.PERIOD_IN_MINUTES * 60) - end[0]) * 1000 : 0;

File: client-agent/src/utils/scheduled-job.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
+const { createSubLogger } = require("../logging");
+const logger = createSubLogger('scheduled-task');
+
+class ScheduledJob {
+    PERIOD_IN_MINUTES;
+    timer;
+    task;
+
+    constructor(func, PERIOD_IN_MINUTES) {
```
Comment (gidimetis) [inline client-agent/src/utils/scheduled-job.js:9]: scheduledFunction ? 

know it already in the context but it helps on search(to me anyway)

**Review by gidimetis — APPROVED** [review]
or maybe after the refactor to typescript, we add name to  each task so 
if an error occured we see it already in the logs for exm:
Error: SLOW_QUERY_LOG TASK FAILED on ..

File: client-agent/src/slow-query-log.js [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+const { QUERIES_FILE } = require('./consts');
+const queriesFileContents = fs.readFileSync(QUERIES_FILE, 'utf8');
+const QUERIES = yaml.load(queriesFileContents);
+const IGNORE_CURRENT_TIME = process.env.IGNORE_CURRENT_TIME === 'true';
+
+const takeAction = async (connectionString, database) => {
+    try {
+        logger.info('takeAction - start');
+        const metisApiKey = consts.API_KEY;
+        const metisExportUrl = consts.API_GATEWAY_HOST;
```
[the remark this answers is not in the document]
Comment (Or92) [reply client-agent/src/slow-query-log.js:18]: irrelevant.
metisExportUrl is the name of the property within the option object.
API_GATEWAY_HOST is being used across the app so i don't want to modify it.

File: client-agent/src/slow-query-log.js [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        logger.info('takeAction - end');
+        return results;
+    }
+    catch (e) {
+        logger.error('error: ', e);
+        return false;
+    }
+}
+
+function getQueries(fakeHoursDelta) {
```
[the remark this answers is not in the document]
Comment (Or92) [reply client-agent/src/slow-query-log.js:43]: this is @liorzmetis function I just used it the way he suggested.

Comment (devopsmetis) [inline client-agent/src/slow-query-log.js:43]: sorry but you dont this function at all....
you can remove it

File: client-agent/src/slow-query-log.js [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    const now = new Date();
+    now.setHours(now.getHours() - fakeHoursDelta);
+    const currentMinutes = now.getMinutes();
+    const currentHour = IGNORE_CURRENT_TIME ? 0 : now.getHours();
+    if (process.argv.length === 2) {
+        return Object.keys(QUERIES)
+            .filter((key) => relevant(QUERIES[key].times_a_day, currentHour, currentMinutes))
+            .map((key) => QUERIES[key]);
+    }
+    const qs = [];
```
[the remark this answers is not in the document]
Comment (devopsmetis) [reply client-agent/src/slow-query-log.js:53]: you can remove all of that, why is it used here ?

File: client-agent/src/slow-query-log.js [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        const nonEligableQueries = process.argv.slice(2).filter((q) => !(q in QUERIES));
+        throw Error(
+            `Error running the CLI. The following are not eligible queries: ${nonEligableQueries}`,
+        );
+    }
+    return qs;
+}
+
+async function collectPlans(fakeHoursDelta, connections) {
+    const theQueries = getQueries(fakeHoursDelta);
```
[the remark this answers is not in the document]
Comment (devopsmetis) [reply client-agent/src/slow-query-log.js:69]: what is it related for? 
you can remove it

File: client-agent/src/slow-query-log.js (hunk 3 above)
Comment (devopsmetis) [inline client-agent/src/slow-query-log.js:81]: change the content as gidi suggest

File: client-agent/src/slow-query-log.js [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+            return takeAction(connectionString, database);
+        }
+    );
+    logger.info('Collection is done.');
+
+    const responses = await Promise.allSettled(promiseArr);
+
+
+    const results = responses
+        .map((responsePromise) => (responsePromise.status === 'fulfilled' && responsePromise.value ? responsePromise.value : []));
```
Comment (devopsmetis) [inline client-agent/src/slow-query-log.js:87]: take care for errors

File: client-agent/src/utils/scheduled-job.js (hunk 1 above)
Comment (devopsmetis) [reply client-agent/src/utils/scheduled-job.js:21]: is it readable? 

@Or92 If so, you can change it to one line

**Pull request merged**